nolearn/lasagne/handlers.py

```diff
@@ -95,6 +95,51 @@ def __call__(self, nn, train_history):
             nn.save_params_to(path)
 
 
+class _RestoreBestWeights:
+    def __init__(self, remember):
+        self.remember = remember
+
+    def __call__(self, nn, train_history):
+        nn.load_params_from(self.remember.best_weights)
+        if self.remember.verbose:
+            print("Loaded best weights from epoch {} where {} was {}".format(
+                self.remember.best_weights_epoch,
+                self.remember.score or self.remember.loss,
+                self.remember.best_weights_loss,
+                ))
+
+
+class RememberBestWeights:
+    def __init__(self, loss='valid_loss', score=None, verbose=1):
+        self.loss = loss
+        self.score = score
+        self.verbose = 1
+        self.best_weights = None
+        self.best_weights_loss = None
+        self.best_weights_epoch = None
+        self.restore = _RestoreBestWeights(self)
+
+    def __call__(self, nn, train_history):
+        key = self.score if self.score is not None else self.loss
+
+        if self.best_weights_loss is None:
+            self.best_weights = nn.get_all_params_values()
+            self.best_weights_loss = train_history[-1][key]
+            self.best_weights_epoch = train_history[-1]['epoch']
+            return
+
+        if self.score is not None:
+            if train_history[-1][key] > self.best_weights_loss:
```

Comment:
then you would negate `train_history[-1][key]` vs `best_weights_loss` and all of the rest of the code would be the same.